airflow/providers/mssql/hooks/mssql_odbc.py

```diff
@@ -0,0 +1,183 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+This module contains MS Sql Odbc hook.
+"""
+
+from urllib.parse import quote_plus
+
+import pyodbc
+
+from airflow.hooks.dbapi_hook import DbApiHook
+
+
+class MsSqlOdbcHook(DbApiHook):
+    """
+    Interact with Microsoft SQL Server using odbc.
+    """
+
+    conn_name_attr = 'mssql_conn_id'
+    default_conn_name = 'mssql_default'
+    supports_autocommit = True
+
+    def __init__(
+        self,
+        mssql_conn_id='mssql_default',
+        database=None,
+        schema=None,
+        driver=None,
+        dsn=None,
+        trx_isolation=pyodbc.SQL_TXN_READ_UNCOMMITTED,
+        autocommit=True,
+        *args,
+        **kwargs,
+    ):
+        super(MsSqlOdbcHook, self).__init__(*args, **kwargs)
+        self.mssql_conn_id = mssql_conn_id
+        self._database = database
+        self._driver = driver
+        self._dsn = dsn
+        self._conn_str = None
+        self.schema = schema
+        self.autocommit = autocommit
+        self._conn_extra = None
+        self._conn_extra_lower = None
+        self._conn = None
+        self.trx_isolation = trx_isolation
+
+    @property
+    def conn(self):
+        """
+        ``airflow.Connection`` object with connection id ``mssql_conn_id``
+        """
+        if not self._conn:
+            self._conn = self.get_connection(self.mssql_conn_id)
+        return self._conn
```

Comment:
```suggestion
        return self._conn if self._conn else self.get_connection(self.mssql_conn_id)
```
I think this will be better, WDYT ;-), so as `conn_extra` and `driver` and `dsn`

Reply:
if you do not store a value to _conn there is no point in using a property

Reply:
Oh, it's a bad review :cry:

Reply:
@zhongjiajie i simplified the `property` situation a bit and tried to remove them where not really beneficial.